askama_derive/src/generator/filter.rs

```diff
@@ -0,0 +1,921 @@
+                name: "escaper",
+                default_value: Some(ARGUMENT_PLACEHOLDER),
+            },
+        ];
+
+        let [source, opt_escaper] = collect_filter_args(ctx, "escape", node, args, ARGUMENTS)?;
+        let opt_escaper = if !is_argument_placeholder(opt_escaper) {
+            let Expr::StrLit(StrLit { prefix, content }) = **opt_escaper else {
+                return Err(ctx.generate_error("invalid escaper type for escape filter", node));
+            };
+            if let Some(prefix) = prefix {
+                let kind = if prefix == StrPrefix::Binary {
+                    "slice"
+                } else {
+                    "CStr"
+                };
+                return Err(ctx.generate_error(
+                    format_args!(
+                        "invalid escaper `b{content:?}`. Expected a string, found a {kind}"
+                    ),
+                    opt_escaper.span(),
+                ));
+            }
+            Some(content)
+        } else {
+            None
+        };
+
+        let escaper = match opt_escaper {
+            Some(name) => self
+                .input
+                .config
+                .escapers
+                .iter()
+                .find_map(|(extensions, path)| {
+                    extensions
+                        .contains(&Cow::Borrowed(name))
+                        .then_some(path.as_ref())
+                })
+                .ok_or_else(|| {
+                    ctx.generate_error(
+                        format_args!(
+                            "invalid escaper '{name}' for `escape` filter. {}",
+                            MsgValidEscapers(&self.input.config.escapers),
+                        ),
+                        node,
+                    )
+                })?,
+            None => self.input.escaper,
+        };
+        buf.write("askama::filters::escape(");
+        self.visit_arg(ctx, buf, source)?;
+        buf.write(format_args!(", {escaper})?"));
+        Ok(DisplayWrap::Wrapped)
+    }
+
+    fn visit_format_filter(
+        &mut self,
+        ctx: &Context<'_>,
+        buf: &mut Buffer,
+        args: &[WithSpan<'a, Expr<'a>>],
+        node: Span<'_>,
+    ) -> Result<DisplayWrap, CompileError> {
+        ensure_filter_has_feature_alloc(ctx, "format", node)?;
+        ensure_no_named_arguments(ctx, "format", args, node)?;
+        if !args.is_empty() {
+            if let Expr::StrLit(ref fmt) = *args[0] {
+                buf.write("askama::helpers::alloc::format!(");
+                self.visit_str_lit(buf, fmt);
+                if args.len() > 1 {
+                    buf.write(',');
+                    self.visit_args(ctx, buf, &args[1..])?;
+                }
+                buf.write(')');
+                return Ok(DisplayWrap::Unwrapped);
+            }
+        }
+        Err(ctx.generate_error(r#"use filter format like `"a={} b={}"|format(a, b)`"#, node))
+    }
+
+    fn visit_fmt_filter(
```

Comment:
Since this is the same code as `format`, shouldn't it be merged into one function and we only pass the filter name for the error messages?

Reply:
It's different from `format`. It's `{{ var | fmt("{:?}") }}` vs `{{ "{var1} {var2}" | format(var1 = 1, var2 = 2) }}`.